### PR TITLE
dynamic validators

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: consensus/istanbul/backends/simple/engine.go [hunk 1]
```diff
@@ -173,10 +173,19 @@ func (sb *simpleBackend) verifyCascadingFields(chain consensus.ChainReader, head
 	if parent.Time.Uint64()+sb.config.BlockPeriod > header.Time.Uint64() {
 		return errInvalidTimestamp
 	}
-	// // Fixed validator set
-	// if bytes.Compare(sb.getValidatorBytes(header), sb.getValidatorBytes(parent)) != 0 {
-	// 	return errInconsistentValidatorSet
-	// }
+	// Retrieve the snapshot needed to verify this header and cache it
+	snap, err := sb.snapshot(chain, number-1, header.ParentHash, []*types.Header{parent})
```
Comment (yutelin) [inline consensus/istanbul/backends/simple/engine.go]: Add command like "Verify validators in extraData. Validators in snapshot and validators in extraData should be the same."

File: consensus/istanbul/backends/simple/engine.go [hunk 2]
```diff
@@ -218,19 +227,25 @@ func (sb *simpleBackend) VerifyUncles(chain consensus.ChainReader, block *types.
...
 	}
 
-	validatorAddresses := validator.ExtractValidators(sb.getValidatorBytes(parent))
-	// ensure the signer is in parent's validator set
-	parentValSet := validator.NewSet(validatorAddresses)
-	if parentValSet == nil {
-		return errInvalidExtraDataFormat
-	}
-	if _, v := parentValSet.GetByAddress(signer); v == nil {
+	if _, v := snap.ValSet.GetByAddress(signer); v == nil {
```
Comment (yutelin) [inline consensus/istanbul/backends/simple/engine.go]: Add comment. Signer should be in the validator set of previous block's extraData.

File: consensus/istanbul/backends/simple/engine.go [hunk 3]
```diff
@@ -298,6 +310,22 @@ func (sb *simpleBackend) Prepare(chain consensus.ChainReader, header *types.Head
 		}
 		sb.candidatesLock.RUnlock()
 	}
+
+	// TODO: we may only insert validator list in epoch block
+	// Ensure the extra data has all it's components
+	snap, err := sb.snapshot(chain, number-1, header.ParentHash, nil)
```
Comment (yutelin) [inline consensus/istanbul/backends/simple/engine.go]: Comment: add validators in snapshot to extraData's validators section.

File: consensus/istanbul/backends/simple/engine.go [hunk 4]
```diff
@@ -258,10 +273,30 @@ func (sb *simpleBackend) Prepare(chain consensus.ChainReader, header *types.Head
 	if !sb.validExtraFormat(parent) {
 		return errInvalidExtraDataFormat
 	}
+	// TODO: we may only insert validator list in epoch block
 	// Ensure the extra data has all it's components
 	header.Extra = sb.prepareExtra(header, parent)
 	// use the same difficulty for all blocks
 	header.Difficulty = defaultDifficulty
+
+	sb.logger.Info("Prepare", "number", number, "sb.config.Epoch", sb.config.Epoch)
```
Comment (yutelin) [inline consensus/istanbul/backends/simple/engine.go]: Comment: add validator voting in coinbase

Comment (yutelin) [inline consensus/istanbul/backends/simple/engine.go]: Also comment that we pick one of the candidates randomly.

File: consensus/istanbul/backends/simple/engine.go [hunk 5]
```diff
@@ -125,20 +122,9 @@ func (sb *simpleBackend) verifyHeader(chain consensus.ChainReader, header *types
...
-			return errInvalidCoinbase
-		}
-		if header.Nonce != (emptyNonce) {
-			return errInvalidNonce
-		}
-	} else {
-		if header.Nonce != (emptyNonce) && !bytes.Equal(header.Nonce[:], nonceAuthVote) && !bytes.Equal(header.Nonce[:], nonceDropVote) {
-			return errInvalidNonce
-		}
+	if header.Nonce != (emptyNonce) && !bytes.Equal(header.Nonce[:], nonceAuthVote) && !bytes.Equal(header.Nonce[:], nonceDropVote) {
```
Comment (yutelin) [inline consensus/istanbul/backends/simple/engine.go]: Update the comment

File: consensus/istanbul/backends/simple/engine.go (hunk 4 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: ok

File: consensus/istanbul/backends/simple/engine.go (hunk 3 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: ok

File: consensus/istanbul/backends/simple/engine.go (hunk 2 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: ok

File: consensus/istanbul/backends/simple/engine.go (hunk 1 above)
Comment (markya0616) [inline consensus/istanbul/backends/simple/engine.go]: ok

File: consensus/istanbul/backends/simple/engine.go [hunk 6]
```diff
@@ -349,17 +426,25 @@ func (sb *simpleBackend) APIs(chain consensus.ChainReader) []rpc.API {
...
+	curHeader := sb.chain.CurrentHeader()
+	snap, err := sb.snapshot(sb.chain, curHeader.Number.Uint64(), curHeader.Hash(), nil)
+	if err != nil {
+		sb.logger.Error("Cannot get latest snapshot", "err", err)
+		return err
+	}
+
+	if _, val := snap.ValSet.GetByAddress(addr); val == nil {
+		sb.logger.Error("Not in validator set", "peerAddr", addr)
+		return errInvalidPeer
```
Comment (tailingchen) [inline consensus/istanbul/backends/simple/engine.go]: should return `istanbul.ErrUnauthorizedAddress`

**Pull request merged**